File: examples/complete.rs
use php_codegen::attribute::AttributeGroup;
use php_codegen::class::Class;
use php_codegen::comment::Document;
use php_codegen::constant::ClassConstant;
use php_codegen::constant::Constant;
use php_codegen::data_type::DataType;
use php_codegen::file::File;
use php_codegen::function::Function;
use php_codegen::interface::Interface;
use php_codegen::literal::Value;
use php_codegen::method::Method;
use php_codegen::modifiers::Modifier;
use php_codegen::modifiers::VisibilityModifier;
use php_codegen::parameter::Parameter;
use php_codegen::property::Property;
use php_codegen::usage::Usage;
use php_codegen::Indentation;

fn main() {
    let file =
        File::new()
            .namespaced("Foo")
            .declare("strict_types", 1)
            .uses("Foo\\Bar\\Baz as Qux")
            .uses("Throwable")
            .uses("DateTime")
            .uses("DateTimeImmutable")
            .uses_function("strlen")
            .uses_function("array_map")
            .uses_function("array_filter")
            .uses_constant("Foo\\HELLO")
            .uses_constant("Foo\\HEY")
            .constant(Constant::new("A").valued("Hello World!"))
            .constant(Constant::new("B").valued(()))
            .constant(Constant::new("C").valued(1))
            .constant(Constant::new("D").valued(false))
            .constant(Constant::new("E"))
            .constant(("F", 213.412))
            .constant(("G", vec![1, 25]))
            .function(
                Function::new("hello")
                    .document(
                        Document::new()
                            .text("This is a simple hello function.")
                            .empty_line()
                            .tag("param", "non-empty-string $firstname")
                            .empty_line()
                            .tag("return", "string")
                            .empty_line()
                            .simple_tag("pure"),
                    )
                    .attributes(AttributeGroup::new().add("Qux", Some("foo: 1, bar: 2")))
                    .parameter(
                        Parameter::new("firstname")
                            .typed(DataType::String)
                            .attributes(
                                AttributeGroup::new()
                                    .add("Validation\\NotBlank", None)
                                    .add("Validation\\Length", Some("min: 2, max: 10")),
                            ),
                    )
                    .parameter(
                        Parameter::new("lastname")
                            .typed(DataType::String)
                            .default(Value::Literal("Qux::Foo".to_string())),
                    )
                    .returns(DataType::String)
                    .body("return 'Hello ' . $firstname . ' ' . $lastname . '!';"),
            )
            .function(Function::new("nothing").returns(DataType::Void))
            .function(
                Function::new("format")
                    .parameter(Parameter::new("template").typed(DataType::String))
                    .parameter(
                        Parameter::new("args")
                            .variadic()
                            .typed(DataType::Union(vec![
                                DataType::Integer,
                                DataType::Float,
                                DataType::String,
                                DataType::Null,
                            ])),
                    )
                    .returns(DataType::String)
                    .body(vec![
                        "return sprintf($template, ...array_map(",
                        "    static fn ($arg) => is_float($arg) ? number_format($arg, 2) : $arg,",
                        "    array_filter($args, static fn ($arg) => $arg !== null)",
                        "));",
                    ]),
            )
            .class(
                Class::new("Example")
                    .extends("Foo\\Bar\\Baz")
                    .implements("Foo\\Bar\\BazInterface")
                    .document(
                        Document::new()
                            .text("This is an example class.")
                            .empty_line()
                            .simple_tag("immutable"),
                    )
                    .modifier(Modifier::Abstract)
                    .using("A")
                    .using(vec!["B", "C"])
                    .using(
                        Usage::new(vec![
                            "D".to_string(),
                            "E".to_string(),
                            "F".to_string(),
                            "G".to_string(),
                        ])
                        .rename("E::bar", "baz")
                        .alias("D::foo", "bar", VisibilityModifier::Public)
                        .public("E::qux")
                        .protected("D::format")
                        .private("D::d")
                        .precede("D::drop", vec!["E"])
                        .precede("G::something", vec!["E", "F", "D"])
                        .visibility("E::e", VisibilityModifier::Protected),
                    )
                    .constant(
                        ClassConstant::new("A")
                            .valued("Hello World!")
                            .modifier(Modifier::Final),
                    )
                    .constant(ClassConstant::new("B").valued(()).protected())
                    .constant(ClassConstant::new("C").valued(1).private())
                    .constant(ClassConstant::new("D").valued(false).public())
                    .property(Property::new("foo").typed(DataType::String).private())
                    .property(Property::new("bar").typed(DataType::String).protected())
                    .property(
                        Property::new("baz")
                            .typed(DataType::Union(vec![DataType::String, DataType::Integer]))
                            .public()
                            .default("Hello World!"),
                    )
                    .method(
                        Method::new("hello")
                            .returns(DataType::String)
                            .parameter(Parameter::new("firstname").typed(DataType::String))
                            .parameter(
                                Parameter::new("lastname")
                                    .typed(DataType::String)
                                    .default(Value::Literal("Qux::Foo".to_string())),
                            )
                            .body("return 'Hello ' . $firstname . ' ' . $lastname . '!';")
                            .attributes(
                                AttributeGroup::new()
                                    .add("Qux", Some("foo: 1, bar: 2"))
                                    .add("Qux", Some("foo: 1, bar: 2")),
                            )
                            .document(
                                Document::new()
                                    .text("This is a simple hello function.")
                                    .empty_line()
                                    .tag("param", "non-empty-string $firstname")
                                    .empty_line()
                                    .tag("return", "string")
                                    .empty_line()
                                    .simple_tag("pure"),
                            ),
                    )
                    .method(
                        Method::new("x")
                            .public()
                            .returns(DataType::Mixed)
                            .body("return 'Hello!';")
                            .attributes(
                                AttributeGroup::new()
                                    .add("Foo", Some("foo: 1, bar: 2"))
                                    .add("Bar", Some("foo: 1, bar: 2")),
                            )
                            .attributes(AttributeGroup::new().add("Baz", None).add("Qux", None))
                            .document(
                                Document::new()
                                    .text("This is a simple x function.")
                                    .empty_line()
                                    .simple_tag("pure"),
                            ),
                    )
                    .method(
                        Method::new("poop")
                            .public()
                            .modifier(Modifier::Abstract)
                            .returns(DataType::Void)
                            .document(Document::new().text("This is a simple poop function.")),
                    )
                    .method(
                        Method::new("helloWorld")
                            .public()
                            .modifier(Modifier::Final)
                            .returns(DataType::Void)
                            .document(Document::new().text("This is a simple echo function."))
                            .body(|indentation: Indentation, level| {
                                indentation.indent("echo 'Hello World!';", level)
                            }),
                    ),
            )
            .interface(
                Interface::new("Formatter")
                    .document(
                        Document::new()
                            .text("This is a simple formatter interface.")
                            .empty_line()
                            .simple_tag("immutable"),
                    )
                    .attributes(
                        AttributeGroup::new()
                            .add("Foo", Some("foo: 1, bar: 2"))
                            .add("Bar", Some("foo: 1, bar: 2")),
                    )
                    .extends("Foo")
                    .extends("Bar")
                    .extends("Qux")
                    .method(
                        Method::new("format")
                            .parameter(Parameter::new("template").typed(DataType::String))
                            .parameter(Parameter::new("args").variadic().typed(DataType::Union(
                                vec![
                                    DataType::Integer,
                                    DataType::Float,
                                    DataType::String,
                                    DataType::Null,
                                ],
                            )))
                            .returns(DataType::String),
                    ),
            );

    print!("{file}");
}

File: tests/complete.php
<?php

declare(strict_types=1);

namespace Foo;

use Foo\Bar\Baz as Qux;
use Throwable;
use DateTime;
use DateTimeImmutable;

use function strlen;
use function array_map;
use function array_filter;

use const Foo\HELLO;
use const Foo\HEY;


const A = "Hello World!";

const B = null;

const C = 1;

const D = false;

const E = null;

const F = 213.412;

const G = [1, 25];

/**
 * This is a simple hello function.
 *
 * @param non-empty-string $firstname
 *
 * @return string
 *
 * @pure
 */
#[Qux(foo: 1, bar: 2)]
function hello(
    #[Validation\NotBlank, Validation\Length(min: 2, max: 10)]
    string $firstname,
    string $lastname = Qux::Foo,
): string {
    return 'Hello ' . $firstname . ' ' . $lastname . '!';
}

function nothing(): void {}

function format(
    string $template,
    int|float|string|null ...$args,
): string {
    return sprintf($template, ...array_map(
        static fn ($arg) => is_float($arg) ? number_format($arg, 2) : $arg,
        array_filter($args, static fn ($arg) => $arg !== null)
    ));
}

/**
 * This is an example class.
 *
 * @immutable
 */
abstract class Example extends Foo\Bar\Baz implements Foo\Bar\BazInterface
{
    use A;
    use B, C;
    use D, E, F, G {
        E::bar as baz;
        D::foo as public bar;
        E::qux as public;
        D::format as protected;
        D::d as private;
        D::drop insteadof E;
        G::something insteadof E, F, D;
        E::e as protected;
    }

    final const A = "Hello World!";

    protected const B = null;

    private const C = 1;

    public const D = false;

    public const bool E = false;

    private string $foo;
    protected string $bar;
    public string|int $baz = "Hello World!";
    /**
     * This is a simple hello function.
     *
     * @param non-empty-string $firstname
     *
     * @return string
     *
     * @pure
     */
    #[Qux(foo: 1, bar: 2), Qux(foo: 1, bar: 2)]
    function hello(
        string $firstname,
        string $lastname = Qux::Foo,
    ): string {
        return 'Hello ' . $firstname . ' ' . $lastname . '!';
    }

    /**
     * This is a simple x function.
     *
     * @pure
     */
    #[Foo(foo: 1, bar: 2), Bar(foo: 1, bar: 2)]
    #[Baz, Qux]
    public function x(): mixed {
        return 'Hello!';
    }

    /**
     * This is a simple poop function.
     */
    public abstract function poop(): void;

    /**
     * This is a simple echo function.
     */
    public final function helloWorld(): void {
        echo 'Hello World!';
    }
}

class SimpleUser
{
    private string $firstName = "Jane";
    private string $lastName = "Doe";
    public string $fullname {
        get {
            return $this->firstName . ' ' . $this->lastName;
        }
        set (string $fullname) {
            [$first, $last] = explode(' ', $fullname);
            $this->firstName = $first;
            $this->lastName = $last;
        }
    }
}

class SimpleUser2
{
    private string $firstName = "Jane";
    private string $lastName = "Doe";
    public string $fullname {
        get {
            return $this->firstName . ' ' . $this->lastName;
        }
    }
}

/**
 * This is an example trait.
 */
trait ExampleTrait
{
    use A;
    use B, C;
    use D, E, F, G {
        E::bar as baz;
        D::foo as public bar;
        E::qux as public;
        D::format as protected;
        D::d as private;
        D::drop insteadof E;
        G::something insteadof E, F, D;
        E::e as protected;
    }

    private string $foo;
    protected string $bar;
    public string|int $baz = "Hello World!";
    /**
     * This is a simple hello function.
     *
     * @param non-empty-string $firstname
     *
     * @return string
     *
     * @pure
     */
    #[Qux(foo: 1, bar: 2), Qux(foo: 1, bar: 2)]
    function hello(
        string $firstname,
        string $lastname = Qux::Foo,
    ): string {
        return 'Hello ' . $firstname . ' ' . $lastname . '!';
    }

    /**
     * This is a simple x function.
     *
     * @pure
     */
    #[Foo(foo: 1, bar: 2), Bar(foo: 1, bar: 2)]
    #[Baz, Qux]
    public function x(): mixed {
        return 'Hello!';
    }

    /**
     * This is a simple poop function.
     */
    public abstract function poop(): void;

    /**
     * This is a simple echo function.
     */
    public final function helloWorld(): void {
        echo 'Hello World!';
    }
}

/**
 * This is an example unit enum.
 */
enum ExampleUnitEnum
{
    use A;
    use B, C;
    use D, E, F, G {
        E::bar as baz;
        D::foo as public bar;
        E::qux as public;
        D::format as protected;
        D::d as private;
        D::drop insteadof E;
        G::something insteadof E, F, D;
        E::e as protected;
    }

    /**
     * This is a foo case.
     */
    case Foo;

    /**
     * This is a bar case.
     */
    case Bar;

    case Baz;
}

/**
 * This is an example string backed enum.
 */
enum ExampleStringBackEnum: string
{
    /**
     * This is a foo case.
     */
    case Foo = "foo value";

    /**
     * This is a bar case.
     */
    case Bar = "bar value";

    case Baz = "baz value";
}

/**
 * This is a simple formatter interface.
 *
 * @immutable
 */
#[Foo(foo: 1, bar: 2), Bar(foo: 1, bar: 2)]
interface Formatter extends Qux
{
    public function format(
        string $template,
        int|float|string|null ...$args,
    ): string;
}
